// @cityofzion/neon-api/dist/provider/dora/class.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dora = void 0;
const neon_core_1 = require("@cityofzion/neon-core");
const core_1 = require("./core");
const log = neon_core_1.logging.default("api");
class Dora {
    constructor(url) {
        if (neon_core_1.settings.networks[url] && neon_core_1.settings.networks[url].extra.dora) {
            this.url = neon_core_1.settings.networks[url].extra.dora;
        }
        else {
            this.url = url;
        }
        log.info(`Created Neoscan Provider: ${this.url}`);
    }
    get name() {
        return `Dora[${this.url}]`;
    }
    getRPCEndpoint() {
        return core_1.getRPCEndpoint(this.url);
    }
    getBalance(address) {
        return core_1.getBalance(this.url, address);
    }
    getClaims(address) {
        return core_1.getClaims(this.url, address);
    }
    getMaxClaimAmount(address) {
        return core_1.getMaxClaimAmount(this.url, address);
    }
    getHeight() {
        return core_1.getHeight(this.url);
    }
    getTransactionHistory(_address) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new Error("Method not implemented.");
        });
    }
}
exports.Dora = Dora;
exports.default = Dora;
